[1단계 - 방탈출 결제/배포] 리버(이연후) 미션 제출합니다.

yeonnhuu opened the pull request:
안녕하세요 파랑 😊
리뷰 잘 부탁드립니다!

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
- https://prolog.techcourse.co.kr/studylogs/4456

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
1. 한 메서드에 오직 한 단계의 들여쓰기만 한다. ✅
2. else 키워드를 쓰지 않는다. ✅
3. 모든 원시값과 문자열을 포장(wrap)한다. ⚠️
4. 한 줄에 점을 하나만 찍는다. ⚠️
5. 줄여쓰지 않는다. ⚠️
6. 모든 entity를 작게 유지한다. ✅
7. 2개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다. ❌
8. 일급 컬렉션을 쓴다. ❌
9. getter/setter/property를 쓰지 않는다. ⚠️

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
외부 API의 응답 지연이나 일시적인 실패 상황에 대한 테스트는 아직 작성하지 못했습니다. 추후 리팩토링 과정에서 해당 테스트를 추가할 예정입니다. 이번 미션에서는 토스 외에도 다양한 외부 API 연동에 유연하게 대응할 수 있도록 설계 분리에 집중했으니, 이 부분 중심으로 리뷰해주시면 감사하겠습니다!
[이번 미션에서의 변경사항](https://github.com/woowacourse/spring-roomescape-payment/pull/218/commits/4c7a128170ca4448536a2ebafb34fc0e7d1cfbe7) 입니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Firedrago95 committed:
feat: migrate previous mission code

Firedrago95 committed:
docs: readme 추가

Firedrago95 committed:
feat: 클라이언트 코드 추가

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
docs: 결제기능 기능목록 작성

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
refactor: dto 필드 추가 및 유효성 검증 로직 추가

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
feat: 예약 결제 기능 구현

Firedrago95 committed:
fix: 예약 결제 요청 로직 수정

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
feat: 예약 결제 요청 예외처리 로직 추가

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
refactor: 예약 결제 요청 예외처리 리팩토링

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: 예약 결제 클라이언트 분리

Firedrago95 committed:
test: 결제 요청 테스트 작성

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
test: 결제요청 통합 테스트 수정

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

Firedrago95 committed:
refactor: payment 패키지 구조 분리

co-authored-by: yeonnhuu <yeonnhuu@gmail.com>

yeonnhuu committed:
refactor: `TossPaymentClient` 설정을 `@Configuration` 기반 Bean 등록 방식으로 분리

- `TossPaymentClient`의 생성과 설정을 `TossPaymentClientConfig`로 분리하여 단일 책임 원칙(SRP) 준수
- RestClient 설정(타임아웃, 인증 헤더 등)을 명확하게 구성 가능하도록 리팩토링

yeonnhuu committed:
feat: 타임아웃 예외 처리 기능 구현

yeonnhuu committed:
refactor: 패키지 구조 개선

summerlunaa commented on src/main/java/roomescape/payment/infrastructure/config/TossPaymentClientConfig.java:
```diff
@@ -0,0 +1,59 @@
+@Configuration
+public class TossPaymentClientConfig {
+
+    private static final String BASE_URL = "https://api.tosspayments.com/v1/payments";
```
참고) Base URL(`https://api.tosspayments.com`)은 secretKey나 timeout 설정처럼 yml이나 properties 등의 설정 파일에 분리하기도 합니다. base url을 분리해두면 환경에 따라 다른 base url 호출이 가능해지거든요. 환경에 따른 유연한 설정이 가능하겠죠?

ex) 실제 운영 환경에서는 `api.tosspayments.com` 호출 / 테스트 환경에서는 `api.beta.tosspayments.com` 호출

그리고 보통 `/v1/payments`는 base url에 포함시키지 않습니다. 현재 설계라면 `/v1/billing/authorizations/issue` 요런 API를 추가로 호출해야 하는 경우 config를 계속 새로 만들어야 겠네요.

yeonnhuu replied:
Base URL(`https://api.tosspayments.com`) 은 application.properties에 분리해두고 `TossPaymentClientConfig`에서는 API 경로(`/v1/payments`)를 조합하는 방식으로 리팩터링하겠습니다!

summerlunaa commented on src/main/java/roomescape/payment/infrastructure/config/TossPaymentClientConfig.java:
```diff
@@ -0,0 +1,59 @@
+    @Value("${payment.toss.secret-key}")
+    private String secretKey;
+
+    @Value("${payment.connect-timeout-length}")
+    private Duration connectTimeout;
```
- Duration 타입으로 받는 이유가 있을까요?
- 테스트 properties에 정의된 값을 사용하고 있네요 🤔

yeonnhuu replied:
- `Duration` 타입을 사용하면 `int`보다 시간 단위를 더 명확하게 표현할 수 있고, Spring Boot의 `@Value` 어노테이션이 이를 자동으로 바인딩해주기 때문에 `toMillis()`와 같은 메서드로 바로 사용할 수 있어 `Duration` 타입을 선택했습니다.
- properties 설정 시 제가 실수했네요...😅 프로덕션 환경용 properties 파일에 정의하도록 수정하겠습니다.

summerlunaa replied:
아하 보통은 초가 아닌 ms로 세팅하긴 합니다. yml 파일에서 시간 단위를 확인할 수 있는 부분은 편하네욥 ㅎㅎ

summerlunaa commented on src/main/java/roomescape/payment/infrastructure/TossPaymentClient.java:
```diff
@@ -0,0 +1,29 @@
+package roomescape.payment.infrastructure;
```
단순 궁금) payment도 reservation과 동일한 레벨의 도메인인가요?

yeonnhuu replied:
미션을 구현할 때 reservation과 payment는 각각 사용자 행위에 대한 독립적인 처리를 담당하는,
동일한 레벨의 도메인으로 접근했습니다. 

사용자가 예약을 먼저 해야 결제를 진행할 수 있기 때문에, 처음에는 결제가 예약에 종속된 것처럼 보일 수도 있어 reservation의 하위 패키지로 두는 것도 고려했습니다. 그러나 현재는 단순히 결제를 승인하는 기능만 포함되어 있어 reservation 하위에 위치해도 큰 문제가 없어 보이지만, 이후 결제 상태 관리, 재시도 로직, 오류 처리 등 기능이 확장될 경우 결제는 예약과 별도의 흐름과 생명 주기를 갖는 도메인이라고 판단했습니다. 즉 reservation과 payment는 서로 연관은 있지만, 동등한 수준의 도메인으로 설계하는 것이 바람직하다고 생각해 payment를 독립된 패키지로 분리했습니다.

summerlunaa replied:
아하 넵넵ㅎㅎ 사실 reservation과 왜 분리했냐는 의도의 질문은 아니었고 현재는 payment 패키지에 domain은 없고 client만 있어서 여쭤봤습니다. 설명 감사합니다.

summerlunaa commented on src/main/java/roomescape/payment/infrastructure/TossPaymentClient.java:
```diff
@@ -0,0 +1,29 @@
+import roomescape.reservation.dto.PaymentRequest;
+import roomescape.payment.dto.PaymentResponse;
+
+public class TossPaymentClient implements PaymentClient {
```
PaymentClient 인터페이스를 두면 어떤 장점이 있나요?

yeonnhuu replied:
`PaymentClient`를 인터페이스로 분리하면 DIP를 따를 수 있어, 서비스와 같은 상위 모듈이 구체적인 구현체에 의존하지 않고 추상화에 의존하게 됩니다. 이로 인해 향후 토스가 아닌 다른 PG사를 연동하는 등의 변경이 발생하더라도 구현체만 교체하면 되고, 서비스 코드는 수정할 필요가 없어 OCP도 만족할 수 있습니다. 또한 이와 같이 인터페이스를 분리한다면 테스트 시 실제 외부 API를 호출하지 않고도 Mock 구현을 주입해 서비스의 동작을 검증할 수 있다는 장점이 있을 것 같습니다.

summerlunaa replied:
네 좋습니다 👍 다만 현재의 PaymentRequest/Response 스펙이 다른 결제사의 client에도 동일하게 적용될 수 있을까는 의문이긴 하네요 🤔

summerlunaa commented on src/main/java/roomescape/payment/infrastructure/TossPaymentClient.java:
```diff
@@ -0,0 +1,29 @@
+
+public class TossPaymentClient implements PaymentClient {
+
+    private final RestClient restClient;
```
RestClient를 사용한 이유가 궁금합니다!

yeonnhuu replied:
이번 미션을 진행하며 학습 자료에 소개된 `RestTemplate`과 `RestClient`에 대해 공부했습니다. `RestTemplate`은 오래된 방식으로, 향후 deprecated될 예정이며 `ClientHttpRequestFactory`나 타임아웃 설정 등 구성 요소가 복잡해 가독성이 떨어지는 단점이 있습니다. 반면 `RestClient`는 `RestTemplate`을 대체하기 위해 도입된 새로운 동기식 HTTP 클라이언트로, `RestClientBuilder`를 활용해 더 깔끔하고 읽기 쉬운 코드를 작성할 수 있고, `onStatus`를 통한 명확한 예외 처리도 가능하다는 장점이 있어, 최종적으로 `RestClient`를 선택했습니다.

summerlunaa commented on src/main/java/roomescape/payment/infrastructure/TossPaymentClient.java:
```diff
@@ -0,0 +1,29 @@
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(paymentRequest)
+                .retrieve()
+                .onStatus(new RequestPaymentErrorHandler(new ObjectMapper()))
```
RequestPaymentErrorHandler와 ObjectMapper를 매번 새로 생성할 필요가 있을까요?

yeonnhuu replied:
`ObjectMapper`를 매번 새로 생성하면 불필요한 객체 생성으로 인해 성능 저하가 발생할 수 있기 때문에 `TossPaymentClient` 내부에서 `ObjectMapper`를 재사용 가능한 인스턴스로 필드에 보관하도록 수정하겠습니다!

summerlunaa commented on src/main/java/roomescape/payment/exception/RequestPaymentErrorHandler.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Override
+    public void handleError(URI url, HttpMethod method, ClientHttpResponse response) throws IOException {
+        String responseBody = new String(response.getBody().readAllBytes(), StandardCharsets.UTF_8);
+        JsonNode errorResponse = objectMapper.readTree(responseBody);
+        String message = errorResponse.get("message").asText();
+        throw new PaymentRequestException(message);
+    }
```
에러 핸들러까지 잘 정의해주셨네요 👍 error code는 기록할 필요 없을까요?

yeonnhuu replied:
토스에서 전달하는 클라이언트 에러 중 일부를 서버 에러로 처리하기 위해, `Error Code`를 기준으로 판단할 필요가 있을 것 같습니다. 이를 위해 서버 에러로 간주할 수 있는 코드를 정의한 `TossErrorCode`를 별도로 추가하겠습니다.

summerlunaa replied:
`TossErrorCode`에는 500으로 처리할 에러 코드만 모아둔 것 같은데 맞을까요? 맞다면 `500으로 처리할 토스 에러 코드`라는 의미를 조금 더 명확하게 알 수 있도록 수정해보면 좋을 것 같아요.

summerlunaa commented on src/main/java/roomescape/reservation/dto/PaymentRequest.java:
```diff
@@ -0,0 +1,8 @@
+        String orderId,
+        int amount
+) {
+}
```
PaymentRequest는 왜 reservation에 있나요?

yeonnhuu replied:
결제 로직을 reservation 패키지에서 payment 패키지로 분리하는 과정에서 실수한 것 같습니다. 수정하도록 하겠습니다!

summerlunaa commented on src/main/java/roomescape/reservation/dto/AdminReservationRequest.java:
```diff
@@ -0,0 +1,15 @@
+    @NotNull(message = "날짜는 null 일 수 없습니다.") LocalDate date,
+    @NotNull(message = "예약 시간 번호는 null 일 수 없습니다.") Long timeId,
+    @NotNull(message = "테마 번호는 null 일 수 없습니다.") Long themeId,
+    @NotNull(message = "페이먼트키는 null 일 수 없습니다.") String paymentKey,
+    @NotNull(message = "주문 번호는 null 일 수 없습니다.") String orderId,
```
문자열인 경우 `@NotEmpty`나 `@NotBlank`도 사용해볼 수 있겠네요.

summerlunaa commented on src/main/java/roomescape/payment/exception/PaymentRequestException.java:
```diff
@@ -0,0 +1,7 @@
+package roomescape.payment.exception;
+
+public class PaymentRequestException extends RuntimeException {
```
PaymentRequestException은 어떤 status code로 처리되나요? 사용자에게는 어떤 예외 메시지가 보여지나요? 개발자는 어떤 이유로 에러가 발생했는지 확인할 수 있나요?

yeonnhuu replied:
현재 `PaymentRequestException`은 별도의 설정이 없어 기본적으로 HTTP 500 서버 오류로 처리되고 있습니다. 토스 API의 응답 JSON에서 `message` 필드 값을 예외 메시지로 그대로 사용하고 있어, 사용자에게 보안상 민감한 정보가 노출될 수 있습니다. 이뿐만 아니라 예외 객체에 메시지만 담겨 있고 에러 코드와 같은 추가 정보가 없어 로그를 통해도 원인을 정확히 파악하기 어렵다는 문제가 있습니다. 

이를 개선하기 위해 `TossPaymentErrorResponse`에 에러 코드와 메시지를 포함하도록 구조를 변경하고, 예외를 `TossPaymentClientException`과 `TossPaymentServerException`으로 구분하겠습니다. 또한 실제로는 토스 API의 클라이언트 오류라 하더라도, 방탈출 서비스 관점에서는 외부 결제 연동 실패로 인해 서버 에러로 처리되도록 개선하겠습니다.

summerlunaa replied:
- `TossPaymentErrorResponse`는 예외 시에만 사용하는 응답 dto네요. exception 보다는 dto 패키지에 있는 게 조금 더 자연스럽겠네요.
- 위에서 `현재의 PaymentRequest/Response 스펙이 다른 결제사의 client에도 동일하게 적용될 수 있을까는 의문`이라고 코멘트 드렸었는데요, 토스의 에러 응답 객체인 `TossPaymentErrorResponse` 처럼 현재의 req, res dto도 사실 toss에만 해당되는 요청/응답 스펙이 아닐까 싶더라구요. 어떻게 생각하시나요?
- 외부 API 호출 시에 발생하는 예외를 호출하는 서버(toss)와 도메인(payment)별로 세분화해서 `TossPaymentXXXException`으로 나눌 필요가 있나요? 외부 API를 연동 할 때마다 예외를 새로 생성해야 할 것 같아서요. 가능하면 커스텀 예외를 조금 더 범용적으로 사용할 수 있게 만들어보면 좋을 것 같아요.
- 신규로 생성된 예외에 대해서도 ExceptionHandler에 추가해야 할 것 같네욥~~

summerlunaa commented on src/test/java/roomescape/reservation/service/ReservationServiceTest.java:
```diff
@@ -260,22 +264,41 @@ public WaitingRepository waitingRepository(JpaWaitingRepository jpaWaitingReposi
+        @Bean
+        public PaymentClient paymentClient() {
+            return new TossPaymentClient(restClient());
+        }
```
현재는 예약 생성 성공 케이스가 없네요. 현재 테스트 코드 상으로는 호출이 안 되는 것 같지만 테스트에서도 tosspay를 실제로 호출하도록 구현한 이유가 궁금합니다 :)

yeonnhuu replied:
예약 생성 기능에 대한 테스트 코드를 추가하겠습니다!

summerlunaa commented on src/test/java/roomescape/payment/infrastructure/TossPaymentClientTest.java:
```diff
@@ -0,0 +1,74 @@
+        .baseUrl(BASE_URL)
+        .defaultStatusHandler(new RequestPaymentErrorHandler(new ObjectMapper()));
+
+    private final MockRestServiceServer server = MockRestServiceServer.bindTo(testBuilder).build();
```
MockRestServiceServer를 이용한 client 테스트를 작성해주셨네요 👍

summerlunaa reviewed (changes requested):
리버 안녕하세요! 파랑입니다 🐳 만나서 반가워요~~
결제 API 연동이 쉽지 않았을 것 같은데 잘 구현해주셨네요 👍 테스트도 잘 챙겨주셨구요!
몇 가지 질문과 리뷰 남겨두었으니 확인 부탁드립니다.
혹시 어떤 주제든 질문이 있거나 리뷰에서 이해 안 되는 부분 있으면 언제든 편하게 DM 주셔도 괜찮습니다~ 마지막 미션도 파이팅 해보시죠 💪

yeonnhuu committed:
:recycle: refactor: 토스페이먼츠 결제 모듈 구조 개선

- `PaymentRequest`를 `payment` 패키지로 이동하여 도메인 경계 명확화
- 예외 처리 클래스 이름을 `TossPayment`로 통일하여 일관성 개선
- `TossPaymentClient`의 예외 처리 로직 강화
- 설정값을 외부 설정으로 분리하여 유연성 향상
- 테스트 코드 수정 및 보완

yeonnhuu committed:
:wrench: chore: 코드 리포맷팅 및 import 최적화

yeonnhuu committed:
:test_tube: test: 예약 생성 테스트 추가

yeonnhuu committed:
:recycle: refactor: 테스트 리팩터링

yeonnhuu commented:
안녕하세요 파랑!
1차 피드백을 반영해서 리팩토링 했습니다.
잘 부탁드립니다 ☺️

### 1차 코드 개선 내용
- 예외 처리 클래스의 네이밍을 `TossPayment`로 통일해 전반적인 일관성과 가독성 향상
- `TossPaymentClient`의 예외 처리 로직을 보완
- 토스 API 오류를 클라이언트 예외(T`ossPaymentClientException`)와 서버 예외(`TossPaymentServerException`)로 구분
- `TossPaymentClientConfig` 내 설정값(예: URL, 시크릿 키 등)을 외부 설정 파일로 분리
- `ReservationServiceTest`에 정상적인 예약 생성 시나리오 테스트 추가
- `TossPaymentClientTest`에 예외 처리 시나리오 테스트 추가
- 토스 클라이언트 오류 중 서버 오류로 간주할 항목을 `TossErrorCode`로 분리
- 토스 API의 에러 응답을 구조화한 `TossPaymentErrorResponse` 클래스를 별도로 분리

summerlunaa reviewed (approved):
리버 안녕하세요!
리뷰도 잘 반영해주시고 남겨드린 질문에도 구체적으로 답변 주셨네요 👍
client에서 발생한 예외를 처리하는 부분에 대해 코멘트 조금 더 남겨드렸는데요, 요 부분은 다음 단계에서 반영해보시면 될 것 같습니다. 1단계는 빠르게 머지하겠습니다. 고생하셨어요 👏
